v2/cmd/asoctl/pkg/importresources/resource_importer.go

```diff
@@ -81,116 +83,149 @@ func (ri *ResourceImporter) Import(
-// would risk deadlock for a sufficiently large resource graph.
-func (ri *ResourceImporter) queueUniqueImporters(
-	candidates <-chan ImportableResource,
-	pending chan<- ImportableResource,
-	progress importreporter.Interface,
-) {
+// This ensures each distinct resource is only imported once,
+func (ri *ResourceImporter) startDeduplicator(
+	resources <-chan ImportableResource,
+	watchdog *watchdog,
+	waitgroup *conc.WaitGroup,
+) chan ImportableResource {
 	seen := set.Make[string]()
 	var queue []ImportableResource
 	var current ImportableResource = nil
 
-	running := true
-	for running {
-		// Dequeue from our internal buffer if needed
-		if current == nil && len(queue) > 0 {
-			current = queue[0]
-			queue = queue[1:]
-		}
+	uniqueResources := make(chan ImportableResource)
 
-		// If we have a current importable to send, use the pending queue
-		var upstream chan<- ImportableResource = nil
-		if current != nil {
-			upstream = pending
-		}
+	waitgroup.Go(func() {
+	run:
+		for {
+			// Dequeue from our internal buffer if needed
+			if current == nil && len(queue) > 0 {
+				current = queue[0]
+				queue = queue[1:]
+			}
 
-		select {
-		case rsrc, ok := <-candidates:
-			if !ok {
-				// Channel closed
-				running = false
-			} else if seen.Contains(rsrc.ID()) {
-				// We've already seen this resource (we've already queued it for import)
-				// So remove it from our count of work to be done
-				ri.log.V(2).Info("Skipping duplicate import", "resource", rsrc.ID())
-				progress.AddPending(-1)
-			} else {
-				// Remember we've seen this, and add it to our queue
-				seen.Add(rsrc.ID())
-				queue = append(queue, rsrc)
-				ri.log.V(2).Info("Buffering import", "resource", rsrc.ID())
+			// If we have a current importable to send, use the pending queue
+			var upstream chan<- ImportableResource = nil
+			if current != nil {
+				upstream = uniqueResources
+			}
+
+			select {
+			case rsrc, ok := <-resources:
+				if !ok {
+					// Channel closed
+					break run
+				} else if seen.Contains(rsrc.ID()) {
+					// We've already seen this resource (we've already queued it for import)
+					// So remove it from our count of work to be done
+					ri.log.V(2).Info("Skipping duplicate import", "resource", rsrc.ID())
+					ri.reporter.AddPending(-1)
+				} else {
+					// Remember we've seen this, and add it to our queue
+					watchdog.starting()
+					seen.Add(rsrc.ID())
+					queue = append(queue, rsrc)
+					ri.log.V(2).Info("Buffering import", "resource", rsrc.ID())
+				}
+			case upstream <- current:
+				// We've sent the current importable, so clear it
+				ri.log.V(2).Info("Queued import", "resource", current.ID())
+				current = nil
 			}
-		case upstream <- current:
-			// We've sent the current importable, so clear it
-			ri.log.V(2).Info("Queued import", "resource", current.ID())
-			current = nil
 		}
+
+		// Close the channel when we're done, so that workers shut down too
+		close(uniqueResources)
```

Comment:
Defer this at the top of the goroutine func to ensure it's closed even in the case of panics?